thread/thread11.h

```diff
@@ -24,160 +25,49 @@ limitations under the License.
-    // value. If a reference argument needs to be passed to the thread function,
-    // it has to be wrapped (e.g., with std::ref or std::cref).
-    template <typename T, typename... Args>
-    struct is_functor
-        : std::conditional<
-              std::is_class<typename std::decay<T>::type>::value &&
-                  std::is_constructible<typename std::function<void(Args...)>,
-                                        typename std::reference_wrapper<
-                                            typename std::remove_reference<
-                                                T>::type>::type>::value,
-              std::true_type, std::false_type>::type {};
-
-    template <typename FUNCTOR, typename... ARGUMENTS>
-    inline typename std::enable_if<is_functor<FUNCTOR, ARGUMENTS...>::value,
-                                   thread*>::type
-    thread_create11(uint64_t stack_size, FUNCTOR&& f, ARGUMENTS&&... args) {
-        // takes `f` as parameter to helper function
-        // thread_create11 will make sure parameters copy is completed
-        return thread_create11(
-            stack_size, &__functor_call_helper<FUNCTOR, ARGUMENTS...>,
-            std::forward<FUNCTOR>(f), std::forward<ARGUMENTS>(args)...);
+// NOTICE: The arguments to the thread function are moved or copied by
+// value. If a reference argument needs to be passed to the thread function,
+// it has to be wrapped (e.g., with std::ref or std::cref).
+template <class F, class... Args, typename = std::enable_if_t<std::__is_invocable<F&&, Args&&...>::value>>
+thread* thread_create11(uint64_t stack_size, F&& f, Args&&... args) {
+    using Gp = std::tuple<typename std::decay<F>::type, typename std::decay<Args>::type...>;
+    std::unique_ptr<Gp> p(new Gp(std::forward<F>(f), std::forward<Args>(args)...));
+    auto thread = thread_create(&__thread_proxy<Gp>, p.get(), stack_size);
+    if (thread) {
+        _unused(p.release());
     }
+    return thread;
+}
 
-    // NOTICE: The arguments to the thread function are moved or copied by
-    // value. If a reference argument needs to be passed to the thread function,
-    // it has to be wrapped (e.g., with std::ref or std::cref).
-    template <typename FUNCTOR, typename... ARGUMENTS>
-    inline typename std::enable_if<is_functor<FUNCTOR, ARGUMENTS...>::value,
-                                   thread*>::type
-    thread_create11(FUNCTOR&& f, ARGUMENTS&&... args) {
-        return thread_create11<FUNCTOR, ARGUMENTS...>(
-            DEFAULT_STACK_SIZE, std::forward<FUNCTOR>(f),
-            std::forward<ARGUMENTS>(args)...);
-    }
+// NOTICE: The arguments to the thread function are moved or copied by
+// value. If a reference argument needs to be passed to the thread function,
+// it has to be wrapped (e.g., with std::ref or std::cref).
+template <class F, class... Args, typename = std::enable_if_t<std::__is_invocable<F&&, Args&&...>::value>>
```

Comment:
`std::__is_invocable` is in same situation as `std::__invoke`

Reply:
> `std::__is_invocable` is in same situation as `std::__invoke`

__is_invocable (std::is_invocable for C++11)

Reply:
![Screenshot_20221205_132725](https://user-images.githubusercontent.com/73515066/205556327-3375f897-3ba6-4826-9ced-beb6ce063545.png)